src/coreclr/jit/lowerarmarch.cpp

```diff
@@ -288,6 +283,63 @@ bool Lowering::IsContainableBinaryOp(GenTree* parentNode, GenTree* childNode) co
         return false;
     }
 
+    if (childNode->OperIs(GT_CAST))
+    {
+        // Find "a op cast(b)"
+        GenTree* castOp = childNode->AsCast()->CastOp();
+
+        // We want to prefer the combined op here over containment of the cast op
+        castOp->ClearContained();
+
+        bool isSupportedCast = false;
+
+        if (varTypeIsSmall(childNode->CastToType()))
+        {
+            // The JIT doesn't track upcasts from small types, instead most types
+            // are tracked as TYP_INT and then we get explicit downcasts to the
+            // desired small type instead.
+
```

Comment:
It would be prudent to add `assert(!varTypeIsFloating(castOp))` here (morph will decompose `double/float -> small type` casts).